### PR TITLE
Add /api/versions endpoint

### PR DESCRIPTION
**Pull request by lukebjerring**
## Description
Lists the versions that we have for a given product. Fairly sure `Distinct` means the sheer quantity of runs (>1000) won't be a problem.

Relates to being consistent with the autocomplete work in #689 

## Review Information
Visit `/api/versions?product=chrome-70`

### CODE REVIEW

File: api/versions.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+		return
+	}
+
+	ctx := shared.NewAppEngineContext(r)
+	query := datastore.NewQuery("TestRun").Filter("BrowserName =", product.BrowserName)
+	queries := []*datastore.Query{query}
+	if product.BrowserVersion != "" {
+		queries = make([]*datastore.Query, 2)
+		queries[0] = query.Filter("BrowserVersion =", product.BrowserVersion).Limit(1)
+		queries[1] = shared.VersionPrefix(query, "BrowserVersion", product.BrowserVersion, false).
```
Comment (mdittmer) [inline api/versions.go]: IIUC, `len(queries) == 1` before this line runs. Should this be `queries = append(queries, [...])`?

Comment (mdittmer) [inline api/versions.go]: nit: Comment reminding meaning of `false` here would make this more readable.

Comment (mdittmer) [inline api/versions.go]: I think `VersionPrefix` is inclusive (`VersionPrefix(foo)` over `["foo", "foobar", "bar"]` yields `["foo", "foobar"]`). Doesn't the inclusion of `"foo"` make `queries[0]` redundant?

File: api/versions.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+			Project("BrowserVersion").
+			Distinct()
+	}
+
+	var runs shared.TestRuns
```
Comment (mdittmer) [inline api/versions.go]: This will initialize to `nil`. Is that safe for the first `append()`, or should we have `runs := make(shared.TestRuns, 0)`?

File: api/versions.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+		w.WriteHeader(http.StatusNotFound)
+		w.Write([]byte("[]"))
+		return
+	}
+
+	versions := make([]string, len(runs))
+	for i := range runs {
+		versions[i] = runs[i].BrowserVersion
+	}
+	sort.Sort(sort.Reverse(sort.StringSlice(versions)))
```
Comment (mdittmer) [inline api/versions.go]: nit: This will put `10.0` before `2.0`, but consistently modeling versions is out of scope here. Maybe add a `TODO`?

File: api/versions.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+	if len(runs) < 1 {
+		w.WriteHeader(http.StatusNotFound)
+		w.Write([]byte("[]"))
+		return
+	}
+
+	versions := make([]string, len(runs))
+	for i := range runs {
+		versions[i] = runs[i].BrowserVersion
```
Comment (mdittmer) [inline api/versions.go]: Follow-up from above: This will probably repeat the one result from `queries[0]` (if there is one) because it will also appear in the result set from `queries[1]`.

File: api/versions.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
+// Copyright 2017 The WPT Dashboard Project. All rights reserved.
```
[the remark this answers is not in the document]
Comment (lukebjerring) [reply api/versions.go]: GET WITH THE TIMES

File: api/versions.go (hunk 1 above)
Comment (lukebjerring) [reply api/versions.go]: You're right and wrong. I don't want the original solo query in the case that a version prefix is needed; I want "exactly that version" and "stuff starting with that version" (which I can't query for both of at once, because, Golang + datastore reasons).

Otherwise the query as built so far is the only one. I'll, uh, make it an else branch for readability.

Comment (lukebjerring) [reply api/versions.go]: Done

Comment (lukebjerring) [reply api/versions.go]: No; because of an edge-case where chrome-7 would match chrome-70, we only find prefixes (and not exact matches).

File: api/versions.go (hunk 2 above)
Comment (lukebjerring) [reply api/versions.go]: `nil` arrays and empty arrays are equal (`len(foo) == 0`, etc)

File: api/versions.go (hunk 3 above)
Comment (lukebjerring) [reply api/versions.go]: Done.

File: api/versions.go (hunk 4 above)
Comment (lukebjerring) [reply api/versions.go]: The tests cover exact matching (and pass) with `chrome-1` :)

https://github.com/web-platform-tests/wpt.fyi/pull/691/files/0086c84e37c341dbed0bedd1e1cf31275285b3d1#diff-8f642c273c44c9c74435d70606cb4148R65

File: api/versions.go (hunk 1 above)
Comment (mdittmer) [reply api/versions.go]: I guess what I don't understand is why isn't it true that "foo" starts with "foo" (because, it does, right?) If it's just something weird about Datastore queries though, then I'm happy to let the tests prove that this is necessary and sufficient.

File: api/versions.go (hunk 2 above)
Comment (mdittmer) [reply api/versions.go]: You're right that append nil-of-type-TestRuns will do what you expect (which is what matters here).

Aside: TIL, though, that an uninitialized TestRuns and make(TestRuns, 0) have the same length, but only the former is equal to nil.

File: api/versions_medium_test.go [hunk 6]
```diff
@@ -33,8 +33,8 @@ func TestApiVersionsHandler(t *testing.T) {
 	json.Unmarshal(bytes, &versions)
 	assert.Equal(t, []string{}, versions)
 
-	// Add test runs
-	someVersions := []string{"2", "1.1.1", "1.1", "1.0", "1"}
+	// Add test runs (duplicating 1.1 is deliberate)
```
Comment (mdittmer) [inline api/versions_medium_test.go]: nit: Comment to end with period.

File: api/versions.go (hunk 1 above)
Comment (lukebjerring) [reply api/versions.go]: Because chrome-7 shouldn't match chrome-70, we cheat by using a prefix of `7.`
This leaves exactly one edge-case; a full string match.

File: api/versions_medium_test.go (hunk 6 above)
Comment (lukebjerring) [reply api/versions_medium_test.go]:
```suggestion
	// Add test runs (duplicating 1.1 is deliberate).
```
